### PR TITLE
Control single- vs multi-group palettes via `col.default`

### LINKED ISSUE #598: Default color for single-group displays: Black vs palette?
As discussed in https://github.com/grantmcdermott/tinyplot/pull/595#issuecomment-4581217133 and also highlighted at the end of https://github.com/grantmcdermott/tinyplot/pull/347#issuecomment-4328439763, we currently do not handle the default color for single-group displays consistently across plot types. Some types use black as the default while others use the first palette color. 

```
tinytheme("clean2")
tinyplot(gear ~ 1, data = mtcars)
tinyplot(factor(gear) ~ 1, data = mtcars)
tinyplot(mpg ~ gear, data = mtcars)
tinyplot(mpg ~ factor(gear), data = mtcars)
```

Image

To resolve this issue we should probably think about the following two questions:

1. Is it sensible to generally use black as the default color for single-group displays?
2. If yes, should the black default be used even if the theme specified a different color in its palette?

### PR DESCRIPTION
**Pull request by grantmcdermott**
Closes #598

@zeileis this PR implements your idea for a new `col.default` parameter that controls the behaviour for single- versus mulii-group displays. I've enacted it for most (but not all) themes, using the following heuristic:

- Turn `col.default = "black`" on for themes that, until now, had a qualitative palette starting with black. This includes the Okabe-Ito-based themes (`"bw"`, `"ipsum2"`, `"broadsheet"`,etc.), plus a few others that conceptually favour a monochrome single series (`"ipsum"`, `"socviz"`, etc.)
  - These themes now drop black from their grouped displays, so a lone series reads as a clean black mark while grouped displays get a full colour palette.
- For "colour-forward" themes (e.g. `"clean(2)"`, `"dark"`, `"web"`), we leave `col.default` as `NULL` so they always use the first colour of their (existing) palette.
- The plain (theme-less) defaults (`"default"`, `"basic"`, etc.) are unchanged: single-group displays fall back to base R `palette()[1]` (i.e., black).

In table form:

  | Theme | `col.default` | Multi-group palette (1st colour) |
  |---|---|---|
  | default / basic / dynamic / tufte | — | base R `palette()` (black) |
  | clean / clean2 | — | Tableau 10 (blue) |
  | dark | — | Set 2 (teal) |
  | nber | — | custom (blue) |
  | web | — | custom (blue) |
  | bw / classic / linedraw / minimal | black | ggplot2 less black (salmon) |
  | ipsum | black | custom (coral) |
  | ipsum2 / broadsheet | black | Okabe-Ito less black (orange) |
  | socviz | black | custom (orange) |
  | ridge / ridge2 | black | Zissou 1 (gradient fills) |
  | float / void | black | "dark" / Tableau 10 |

### CODE REVIEW

File: R/type_histogram.R [hunk 1]
```diff
@@ -106,7 +106,6 @@ data_histogram = function(breaks = "Sturges",
         hbreaks = ifelse(!sapply(.breaks, is.null), .breaks, "Sturges")
 
         if (is.null(by) && is.null(palette)) {
```
[the remark this answers is not in the document]
Comment (grantmcdermott) [reply R/type_histogram.R]: This is a pre-existing "bug". Let me see how easy it is to fix here...

**Comment by grantmcdermott** [comment]
P.S. Here are some example, using your four plot case from #598 

First, the default (no) theme:
``` r
pkgload::load_all("~/Documents/Projects/tinyplot/")
#> ℹ Loading tinyplot
par(mfrow = c(2,2))
tinyplot(gear ~ 1, data = mtcars)
tinyplot(factor(gear) ~ 1, data = mtcars)
tinyplot(mpg ~ gear, data = mtcars)
tinyplot(mpg ~ factor(gear), data = mtcars)
```



Second, using `"clean2"` theme.

``` r
tinytheme("clean2")
tinyplot(gear ~ 1, data = mtcars)
tinyplot(factor(gear) ~ 1, data = mtcars)
tinyplot(mpg ~ gear, data = mtcars)
tinyplot(mpg ~ factor(gear), data = mtcars)
```



Third, using `"ipsum"` theme.

``` r
tinytheme("ipsum")
tinyplot(gear ~ 1, data = mtcars)
tinyplot(factor(gear) ~ 1, data = mtcars)
tinyplot(mpg ~ gear, data = mtcars)
tinyplot(mpg ~ factor(gear), data = mtcars)
```

File: R/type_histogram.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (grantmcdermott) [reply R/type_histogram.R]: https://github.com/grantmcdermott/tinyplot/pull/614/commits/9a3a6c6bb3c1b228c58bf35943b72ac5a2fea23e

I tweaked the code so that we drop the (unused) neutral gray branch and rather just use stick with `ribbon.alpha` transparency for histogram fill. I could have fixed the conditional logic to the former, but then we would have had an inconsistency between the alpha fill of different themes. The visual difference is basically indistinguishable anyway.

**Comment by grantmcdermott** [comment]
@vincentarelbundock I realized that you might to take a look at this too, since it will effect behaviour for some themes.

**Comment by vincentarelbundock** [comment]
oh yeah, that seems like a good idea to me! I like the examples a lot.

**Comment by zeileis** [comment]
Thanks, Grant @grantmcdermott! Sorry for the slow follow-up...the last weeks were very busy with the preparations for our [World Cup forecast](https://www.zeileis.org/news/fifa2026/) and corresponding press work.

Regarding the `col.default`:

- Using a `col.default` that is not part of `palette.qualitative` works great.
- As does re-using the first palette color.
- But using the first color and dropping it in the multi-group case should be facilitated.
- I will have a closer look at this and try to come up with a suggestion.

Regarding plots with shaded areas such as `type_histogram`, `type_boxplot`, `type_barplot`.

- For shaded areas we should use a lighter version of `col.default` by default.
- `type_histogram` and `type_boxplot` do so albeit via alpha blending.
- `type_barplot` uses the `col.default` directly which will typically be rather "heavy".
- I think it would be good to use a unified approach here that uses a lighter (but still opaque) version of `col.default` for shading areas.
- The trick via alpha blending is nice as long as there are no grid lines. With grid lines, opaque colors would work better.
- I'm not sure what is the best place in the code to resolve this.

**Comment by zeileis** [comment]
I think it would be nice to support `col.default` specifications relative to `palette.qualitative`, e.g.,

```
col.default = -1, palette.qualitative = "Okabe-Ito"
```

which should mean: use the first color from the palette as the default color and take it out of the qualitative palette afterwards. Then we don't need to keep around extra copies of palettes without black.

In order to resolve this we would have to add something like the following code somewhere in the `tpar` computations:

```
if (is.character(palette.qualitative) && length(palette.qualitative) == 1L) {
  idx = match_palette_name(palette.qualitative, palette.pals())
  if (!is.na(idx) && idx >= 1L) palette.qualitative = palette.colors(palette = palette.pals()[idx])
}
if (is.character(palette.qualitative) && length(palette.qualitative) == 1L) {
  idx = match_palette_name(palette.qualitative, hcl.pals())
  if (!is.na(idx) && idx >= 1L) palette.qualitative = hcl.colors(palette = hcl.pals()[idx])
}
if (is.numeric(col.default) && length(col.default) == 1L) {
  idx = col.default
  col.default = palette.qualitative[abs(idx)]
  if (idx < 0L) palette.qualitative = palette.qualitative[idx]
}
```

**Comment by grantmcdermott** [comment]
Thanks for the suggestions and review @zeileis. I think I've manged to implement your core _desiderata_. For example, you can now use `col.default = -1` (and we use this pattern internally):

``` r
pkgload::load_all("~/Documents/Projects/tinyplot/")
#> ℹ Loading tinyplot
tinytheme_register(
  "mytheme", theme = "default",
  col.default = -1, palette.qualitative = "Okabe-Ito"
)
tinyplot(Sepal.Length ~ Petal.Length | Species, iris, theme = "mytheme")
```




Similarly, I've reworked the "fill" logic for the `type_histogram`, `type_boxplot`, `type_barplot`, and `type_ribbon`. The _tl;dr_ version is that these types now all (re-)use the same `seq_palette()` logic that you originally introduced for `type_spineplot`. This removes the alpha blending and shading inconsistency that you highlighted, and ensures better fill matching, both across types and w.r.t. base R plots. (The one exception is that `type_barplot()` now reverts to "lightgray" filling versus "gray", which is what base `barplot()` does. But I felt internal consistency was the right thing to prioritize.)

I think you get a decent sense of the improvement via our running four-plot example:

``` r
op = par(mfrow = c(2,2))
tinyplot(gear ~ 1, data = mtcars, grid = TRUE)
tinyplot(factor(gear) ~ 1, data = mtcars, grid = TRUE)
tinyplot(mpg ~ gear, data = mtcars)
tinyplot(mpg ~ factor(gear), data = mtcars)
```



``` r
tinytheme("clean2")
tinyplot(gear ~ 1, data = mtcars)
tinyplot(factor(gear) ~ 1, data = mtcars)
tinyplot(mpg ~ gear, data = mtcars)
tinyplot(mpg ~ factor(gear), data = mtcars)
```



``` r
tinytheme()
tinytheme("ipsum")
tinyplot(gear ~ 1, data = mtcars)
tinyplot(factor(gear) ~ 1, data = mtcars)
tinyplot(mpg ~ gear, data = mtcars)
tinyplot(mpg ~ factor(gear), data = mtcars)
```



``` r
tinytheme()
par(op)
```

<sup>Created on 2026-06-15 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Comment by zeileis** [comment]
Grant, cool, thanks a lot! The new examples look really nice. :partying_face: 

Should I try to do another round of playing around with the new features to see if I find anything else? (Can't do so today but will do in the next days.)

**Comment by grantmcdermott** [comment]
Thanks @zeileis.

> Should I try to do another round of playing around with the new features to see if I find anything else?

Normally, I would say "yes", but I'm a little anxious to get this rolled in so that I can start closing the other remaining loose ends before 0.7.0 CRAN submission (which I'm really hoping to send by EOW). I've done my best to test a bunch of edge case, so feel pretty confident... But please kick the tyres some more when you get a chance. We can also add additional fixes if we spot anything down the line.

Thanks again for all of your great suggestions, both on the design side and the outcome side. It made things _much_ easier once I had something. principled to work towards.

P.S. Here's one last example: a "manual" version of the pairs-esque plot that you've suggested in #613 (which I plan to look at tomorrow).

```r
pkgload::load_all("~/Documents/Projects/tinyplot/")
png('tmp.png', width = 16, height = 16, units = "in", res = 150)
op = par(mfrow = c(5, 5))
tinytheme('clean2')
for (i in names(iris)) {
  for(j in names(iris)) {
    dat = iris[, unique(c(j, i)), drop = FALSE]
    if (identical(i, j)) {
      fml = reformulate(i)
      typ = if (inherits(dat[[i]], "numeric")) "hist" else NULL
    } else {
      fml = reformulate(j, response = i)
      typ = NULL
    }
    tinyplot(fml, dat, type = typ)
    # box('figure', lwd = 0.5)
  }
}
par(op)
dev.off()
```

tmp

**Comment by zeileis** [comment]
Fair enough, Grant, thanks!

Thanks also for the updated pairs-esque plot. I think you are right in #613 and we should add it to `tinyplot.data.frame()` even if it is not perfect, yet. Maybe we can play around a little bit with the axis labels and only plot them on the bottom row and left column.

Another thing I noticed is that `pch = "by"` does not use the `pch` variable from the theme (if any). Is something for a separate issue - or is there another way of doing it? Consider the following example:

```
tinytheme(col.default = -1, palette.qualitative = "R4", pch = 15:17)
tinyplot(bill_dep ~ bill_len | species, data = penguins, pch = "by")
tinyplot(bill_dep ~ bill_len | species, data = penguins, pch = 15:17)
```

tinyplot-pch

Is there another way to make the plot from `pch = "by"` look like `pch = 15:17`?

**Comment by grantmcdermott** [comment]
> Another thing I noticed is that pch = "by" does not use the pch variable from the theme (if any). Is something for a separate issue - or is there another way of doing it?

Hmmm. I would have said that the "correct" way to do it is simply registering the starting `pch` value in the theme and let it recycle from there. But that doesn't work either.

```r
tinytheme(col.default = -1, palette.qualitative = "R4", pch = 15)
tinyplot(bill_dep ~ bill_len | species, data = penguins, pch = "by")
```

So, definitely a bug. Do you mind filing a separate issue?
